### PR TITLE
feat: code struct init

### PR DESCRIPTION
**Pull request by VaderKai**
#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.
- [x] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en:
zh(optional):

### CODE REVIEW

File: pkg/data/null.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package data
+
+import commons "github.com/kitex-contrib/codec-hessian2/pkg/common"
```
Comment (felix021) [inline pkg/data/null.go]: no need to rename the package here

File: pkg/data/null.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+// EncNull ::= 'N'
+func EncNull(i interface{}) byte {
```
Comment (felix021) [inline pkg/data/null.go]: input param is not needed here

Comment (felix021) [reply pkg/data/null.go]: Don't use abbriviation for better readability

Comment (felix021) [reply pkg/data/null.go]: I think these APIs will be better:
- `EncodeNull(w io.Writer)` 
- `DecodeNull(tag byte, r io.Reader)` (tag is already read by the caller)

File: pkg/decoder.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	}
+}
+
+type Decoder struct {
+	reader io.Reader     // input stream
+	buffer *bytes.Buffer // buffer cache
+}
+
+func (e *Decoder) Decode(obj interface{}) error {
+	// nil or nil pointer
```
Comment (felix021) [inline pkg/decoder.go]: Not sure if it's OK to deal with nil here. Maybe we should deal it only for map/list/objects. We can fix it later if so.

File: pkg/decoder.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+func NewHessian2Input(r io.Reader) *Decoder {
+	return &Decoder{
+		reader: r,
+		buffer: bytes.NewBuffer(nil),
+	}
+}
+
+type Decoder struct {
```
Comment (felix021) [inline pkg/decoder.go:36]: no need to separate encoder/decoder.
pls refer to thriftCodec 
 https://github.com/cloudwego/kitex/blob/develop/pkg/remote/codec/thrift/thrift.go#L67

File: pkg/payload_codec.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+type ByteBuffer struct {
+}
+
+// Hessian2Codec NewHessian2Codec creates the hessian2 codec.
+type Hessian2Codec struct {
+}
+
+// NewHessian2Codec creates a new codec instance.
+func NewHessian2Codec() *Hessian2Codec {
+	return nil
```
Comment (felix021) [inline pkg/payload_codec.go]:
```
// TOOD: WIP
panic("not implemented")
```

same below in Marshal and Unmarshal.

File: pkg/data/null.go (hunk 2 above)
Comment (jasondeng1997) [reply pkg/data/null.go]: 这个包有问题

File: pkg/decoder.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+package hessian2
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+func NewHessian2Input(r io.Reader) *Decoder {
```
Comment (tiltwind) [inline pkg/decoder.go]: `NewHessian2Input`, `NewHessian2Output` are java style naming,  how about change to golang style `NewDecoder`, `NewEncoder`?

File: pkg/data/null.go (hunk 1 above)
Comment (VaderKai) [reply pkg/data/null.go]: done

File: pkg/data/null.go (hunk 2 above)
Comment (VaderKai) [reply pkg/data/null.go]: update to bufio.writer/reader

File: pkg/decoder.go (hunk 3 above)
Comment (VaderKai) [reply pkg/decoder.go]: ok

File: pkg/payload_codec.go (hunk 5 above)
Comment (VaderKai) [reply pkg/payload_codec.go]: done

File: pkg/decoder.go (hunk 6 above)
Comment (VaderKai) [reply pkg/decoder.go]: done

**Review by jasondeng1997 — APPROVED** [review]

**Comment by jasondeng1997** [comment]
LGTM

File: pkg/common/type.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package commons
+
+// Type constants in the Thrift protocol
```
Comment (felix021) [inline pkg/common/type.go]: `Hessian2`, not `Thrift`

File: pkg/hessian_protocol.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	"sync"
+)
+
+// must be strict read & strict write
+var (
+	bpPool sync.Pool
+	_      BaseProtocol = (*BinaryProtocol)(nil)
+)
+
+func newBP() interface{} {
```
Comment (felix021) [inline pkg/hessian_protocol.go:38]: Not referenced?
You might need to add an `init()` with `bpPool.New = newBP`;

File: pkg/struct/base.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package hessian2
```
Comment (felix021) [inline pkg/struct/base.go]: What is this `Base` for?

Comment (VaderKai) [reply pkg/struct/base.go]: done

File: pkg/common/type.go (hunk 7 above)
Comment (VaderKai) [reply pkg/common/type.go]: done

File: pkg/hessian_protocol.go (hunk 8 above)
Comment (VaderKai) [reply pkg/hessian_protocol.go:38]: done

**Review by felix021 — APPROVED** [review]

**Pull request merged**